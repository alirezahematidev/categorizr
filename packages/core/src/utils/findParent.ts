import { TreeNode } from "$core/types";
import { noChildren } from "./helpers";

export function findParent<T extends TreeNode>(tree: T[], node: T): T | undefined {
  for (const branch of tree) {
    if (noChildren(branch)) continue;

    if (branch.children.some((child) => child.id === node.id)) return branch;

    const parent = findParent<T>(branch.children as T[], node);

    if (parent) return parent;
  }

  return undefined;
}
