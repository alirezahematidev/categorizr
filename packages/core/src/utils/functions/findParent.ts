import { MaybeNode, TreeNode } from "$core/types";
import { noChildren } from "../helpers";

export function findParent<T extends TreeNode>(tree: readonly T[], node: T): MaybeNode<T> {
  for (const branch of tree) {
    if (noChildren(branch)) continue;

    if (branch.children.some((child) => child.id === node.id)) return branch;

    const parent = findParent<T>(branch.children as T[], node);

    if (parent) return parent;
  }

  return undefined;
}
