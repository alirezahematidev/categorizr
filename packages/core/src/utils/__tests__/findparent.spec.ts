import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { findParent } from "../findParent";
import type { TreeNode } from "$core/types";

describe("findParent", async () => {
  beforeAll(() => {
    vi.fn().mockImplementation(findParent);
  });

  afterAll(() => {
    vi.resetAllMocks();
  });

  const { default: CATEGORIES } = await vi.importActual<{ default: TreeNode[] }>("$core/__mocks__");

  it("returns undefined for empty tree", () => {
    const emptyTree: TreeNode[] = [];

    const node = {
      id: "1",
      name: "category-1",
      children: [],
    };

    expect(findParent(emptyTree, node)).toBeUndefined();
  });

  it("returns undefined when node is not found", async () => {
    const node = {
      id: "100",
      name: "non-existent-node",
      children: [],
    };

    expect(findParent(CATEGORIES, node)).toBeUndefined();
  });

  it("returns correct parent node for node with children", async () => {
    const node = {
      id: "3",
      name: "sub-category-1",
      children: [],
    };

    expect(findParent(CATEGORIES, node)).toMatchSnapshot();
  });

  it("returns correct parent node for node with children", async () => {
    const node = {
      id: "4",
      name: "sub-category-1",
      children: [],
    };

    expect(findParent(CATEGORIES, node)).toMatchSnapshot();
  });
});
