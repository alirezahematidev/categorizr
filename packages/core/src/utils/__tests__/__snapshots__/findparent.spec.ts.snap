// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`findParent > returns correct parent node for node with children 1`] = `
{
  "children": [
    {
      "children": [],
      "id": "3",
      "name": "sub-category-1",
    },
  ],
  "id": "1",
  "name": "category-1",
}
`;

exports[`findParent > returns correct parent node for node with children 2`] = `
{
  "children": [
    {
      "children": [],
      "id": "4",
      "name": "sub-category-2",
    },
  ],
  "id": "2",
  "name": "category-2",
}
`;
